dhall-docs/src/Dhall/Docs/Core.hs

```diff
@@ -223,90 +266,132 @@ saveHtml inputAbsDir outputAbsDir packageName DhallFile {path = absFile, ..} = d
       where
         strippedHeader = Data.Text.strip h
 
+{-| Create an @index.html@ file on each available folder in the input.
 
-{-| Create an index.html file on each folder available in the second argument
-    that lists all the contents on that folder.
+    Each @index.html@ lists the files and directories of its directory. Listed
+    directories will be compacted as much as it cans to improve readability.
 
-    For example,
+    For example, take the following directory-tree structure
 
-    @
-    createIndexes [absdir|/|]
-        [ [absfile|\/a\/b.txt|]
-        , [absfile|\/a\/c/b.txt|]
-        , [absfile|\/a\/c.txt"|]
-        ]
-    @
+    > .
+    > ├── a
+    > │   └── b
+    > │       └── c
+    > │           └── b.dhall
+    > └── a.dhall
 
-    ... will create two index.html files:
+    To improve navigation, the index at @./index.html@ should list
+    @a/b/c@ and no @index.html@ should be generated inside of `a/` or
+    `a/b/`, but yes on `a/b/c/` in the last one there is the @b.dhall@ file
+-}
+createIndexes :: Text -> [DhallFile] -> [(Path Rel File, Text)]
+createIndexes packageName files = map toIndex dirToDirsAndFilesMapAssocs
+  where
+    -- Files grouped by their directory
+    dirToFilesMap :: Map (Path Rel Dir) [DhallFile]
+    dirToFilesMap = Map.unionsWith (<>) $ map toMap files
+      where
+        toMap :: DhallFile -> Map (Path Rel Dir) [DhallFile]
+        toMap dhallFile =
+            Map.singleton (Path.parent $ path dhallFile) [dhallFile]
+
+    {-  This is used to compute the list of exported packages on each folder.
+        We try to compress the folders as much as we can. See `createIndexes`
+        documentation to get more information.
+    -}
+    dirToDirsMap :: Map (Path Rel Dir) [Path Rel Dir]
+    dirToDirsMap = Map.map removeHereDir $ foldl go initialMap dirs
```

Comment:
In general, I'd recommend using `foldl'` over `foldl` for folding over lists. There's a bit of a discussion here: https://github.com/hasura/graphql-engine/pull/2933#discussion_r328821960